### PR TITLE
Add MultiDeviceTestContext for testing more than one Device

### PR DESCRIPTION
**Pull request by reszelaz**
First of all thanks to @ajoubertza and @tiagocoutinho for your suggestions to the interface.  Also, if you find that the proposed interface could be improved just let me know and I will change it.

DeviceTestContext runs just one Device hence is only useful for unit testing of one Device. Integration tests involving multiple Devices within the same server are not possible. Add MultiDeviceTestContext for that purpose. This context thanks to a little bit more complicated interface allows to run multiple Devices (each of them even exporting multiple devices) all within one server.

At the same time refactor the DeviceTestContext to use the MultipleDeviceTestContext as its base class. This way, the DeviceTestContext is just a simplified interface to its base class.

### CODE REVIEW

**Comment by reszelaz** [comment]
An example of usage:

```python
from tango import DeviceProxy
from tango.test_context import MultiDeviceTestContext

from sardana.tango.pool.Pool import Pool, PoolClass
from sardana.tango.pool.Controller import Controller, ControllerClass


devices_info = (
    {
        "class": (PoolClass, Pool),
        "devices": (
            {
                "name": "pool/test/1"
            },
        )
    },
    {
        "class": (ControllerClass, Controller),
        "devices": (
            {
                "name": "controller/test/1",
                "properties": {
                   "Id": 1,
                   "Klass": "DummyMotorController",
                   "Library": "DummyMotorController.py",
                   "Type": "Motor"
                }
            },
        )
    }
)

def test_controller():
    """Test device voltage reading after calibration."""
    with MultiDeviceTestContext(devices_info, process=True, host="localhost") as context:
        pool = context.get_device("pool/test/1")
        print(pool.ControllerList)
```

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Interface looks good.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py:282]: Looking at this, shouldn't we move `device_cls  = None`  (L163) inside the loop?  We might have input like this, where later devices don't specify a class:
```python
devices_info = (
    {
        "class": (PoolClass, Pool),
        "devices": (
            {
                "name": "pool/test/1"
            },
        )
    },
    {
        "class": MyDevice,
        "devices": (
            {
                "name": "my/device/1"
            },
        )
    }
)
```

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: `server_name` and `instance_name` don't change during the loop, while `tangoclass` will.  I'm guessing it will be a problem if someone specifies a config that results in the same `tangoclass` more than once.  Will it? If so, should we do some validation, or will the problem be obvious from the failure?

Alternatively, could we make `tangoclass` unique, or must it be `device.__name__`?

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py:297]: Maybe we should warn or fail if both `class_list` and `device_list` are not empty.  In that case, some of the devices requested by the user will not be started up.  It looks like that combination is not supported.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: I think we can remove this method.  It isn't used anymore.  I can't see why someone would want to use it externally to the `DeviceTestContext` initialisation either.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Is the `flush` necessary, given that exiting the `open` context handler closes the file?  Same for the next method.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: If there are multiple devices we'd end up with a file like this:
```
Clock/clock/DEVICE/Clock: "test/nodb/clock1"
: "test/nodb/clock2"
```
Will that work?  I'm not sure about the format of these files.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Suggest:  `"""Return new device proxy ..."""`
Important to know that a new instance is being instantiated by calling this method.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Suggest:   `# Single device test context`

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: If we remove the old `generate_db_file` method, then I think we might as well combine this method with `generate_db_file_device`.   I don't like that `generate_db_file_tangoclass` leaves the file in an incomplete state.

New combined method can then be called `generate_db_file` or maybe `update_db_file` or `append_db_file`.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py:255]: I wonder if not waiting to ping all the devices might cause some problems when using this for unit tests.  Maybe the server is up but not all the devices.  What about making a dict of `DeviceProxy` instances for all the devices, and pinging them here?   Then we wouldn't need to make a new proxy on each call to `get_device()`.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py:430]: This is a little different to the `DeviceTestContext`.  We could return the dict of device proxies here.  Although, sometimes it is useful to have access to the whole object (e.g. to access the server, or get the device access URL).  I don't mind, either way.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Think this can be removed (maybe left over from debugging?)

**Review by ajoubertza — CHANGES_REQUESTED** [review]
Very nice, @reszelaz!  Thanks for the improvement.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:282]: You are right! I'm correcting this. Thanks for spotting that!

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Yes, configuration resulting in repeating the tangoclass fail as it is now. Furthermore the error message is not intuitive at all. I will add some validation.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:297]: You are right! I'm adding this validation. IMO warn is too soft, I would raise a `ValueError`.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Ok, I left it just in case only for backwards compat. I'm removing it.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Ok, I merge these two methods `generate_db_file_tangoclass` and `generate_db_file_device` into one:
```python
def append_db_file(self, server, instance, tangoclass, device_prop_info):
    [...]
```

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:255]: Sounds good. I'm proceeding with the change.

Comment (reszelaz) [reply tango/test_context.py:255]: Actually, with the tests that I did, I observe that the ping to dserver device replies only after all the other devices are exported. Another note is that the ping is not "waiting". If there is no connection, the ping simply raises an exception. 

IMO, ping to all the others devices is "just in case". Do you know any example of what could go wrong so the other devices does not respond to ping and the dserver responds?

I imagine that in Sardana we will make a generic test context that will populate "sar_demo" device server with approx. 30 devices. Then in particular tests we will just use proxies to, let's say, a couple of them. I think that preparing all the 30 device proxies up-front is not optimal. What I see good is to maintain cache of the device proxies within the context so it will be filled gradually on calls to `get_device`. Or simple don't maintain this cache and leave it to the test developer to keep them and not call `get_device` every time it is necessary to do something with the device.

I'll wait with implementing these changes for your opinion.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:430]: IMHO self is enough. Anyway it is somehow related to the previous comment (whether we create device proxies upfront).

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Sure! I'm removing this.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Ok

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:282]: Done in b3c6316.
But actually this context is not yet ready for mixing both: HLAPI and classical API.
If someone wants in the future to enhance it this part will be already done:)

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in: f6286e3

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:297]: Done in: e693575

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in:  0764edb

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py]: OK.   Thanks for including the commit hash - very easy to review

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py:255]: > Actually, with the tests that I did, I observe that the ping to dserver device replies only after all the other devices are exported. Another note is that the ping is not "waiting". If there is no connection, the ping simply raises an exception.

OK, good to know.  OK, I see, the ping will not wait.

> IMO, ping to all the others devices is "just in case". Do you know any example of what could go wrong so the other devices does not respond to ping and the dserver responds?

No, no examples.

> I imagine that in Sardana we will make a generic test context that will populate "sar_demo" device server with approx. 30 devices. Then in particular tests we will just use proxies to, let's say, a couple of them. I think that preparing all the 30 device proxies up-front is not optimal. What I see good is to maintain cache of the device proxies within the context so it will be filled gradually on calls to `get_device`. Or simple don't maintain this cache and leave it to the test developer to keep them and not call `get_device` every time it is necessary to do something with the device. 

That's an interesting use case.  With a little effort, we could get the best of both.  I'm thinking we could create a dict-like object that lazy loads the device proxies.  We return this to the user (or they access it via a `self.devices` property).  The first time a key is accessed, the `DeviceProxy` object is created.  What do you think?

Comment (reszelaz) [reply tango/test_context.py:255]: Done in 0dac6a9.

Finally, to make it simpler, I went for an option for directly keeping the dictionary within the context class scope. A dedicated class for this purpose would need to know the host, port, nodb to be able to compose the Tango URI. This is because we want to get the devices using just the family/domain/member names, as these are used in the `devices_info` argument.

Comment (ajoubertza) [reply tango/test_context.py:255]: OK.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: We should make use of `get_device` here so that the proxy object is in the cache.  That has the added benefit that accessing the proxy via `self.device` or `self.get_device` returns the same instance.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: I think we should remove this docstring, and put all the info in the class docstring.  That is what has been done in other parts of the code.  E.g. https://github.com/tango-controls/pytango/blob/492c8752650744b88112bebc450f8497a8b0315b/tango/server.py#L1183-L1204

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py:215]: I think we should add some more detail here.  The example is sardana, but maybe that should be changed to some trivial example, so users don't have to install sardana to try this out?

```suggestion
    """
    Example usage::

        from tango.test_context import MultiDeviceTestContext

        from sardana.tango.pool.Pool import Pool, PoolClass
        from sardana.tango.pool.Controller import Controller, ControllerClass

        devices_info = (
            {
                "class": (PoolClass, Pool),
                "devices": (
                    {
                        "name": "pool/test/1"
                    },
                )
            },
            {
                "class": (ControllerClass, Controller),
                "devices": (
                    {
                        "name": "controller/test/1",
                        "properties": {
                           "Id": 1,
                           "Klass": "DummyMotorController",
                           "Library": "DummyMotorController.py",
                           "Type": "Motor"
                        }
                    },
                    {
                        "name": "controller/test/2",
                        "properties": {
                           "Id": 2,
                           "Klass": "DummyMotorController",
                           "Library": "DummyMotorController.py",
                           "Type": "Motor"
                        }
                    },
                )
            }
        )

        def test_controller():
            with MultiDeviceTestContext(devices_info, process=True) as context:
                pool = context.get_device("pool/test/1")
                print(pool.ControllerList)

    :param devices_info:
      a sequence of dicts with information about
      devices to be exported. Each dict consists of the following keys:
        * "class" which value is either of:
          * :class:`~tango.server.Device`
          * a sequence of two elements :class:`~tango.DeviceClass`
            and :class:`~tango.DeviceImpl`
        * "devices" which value is a sequence of dicts with the following keys:
          * "name" (str)
          * "properties" (dict)
    :type devices_info:
      sequence<dict>
    :param server_name:
      Name to use for the device server.
      Optional.  Default is the first device's class name.
    :type server_name:
      :py:obj:`str`
    :param instance_name:
      Name to use for the device server instance.
      Optional.  Default is lower-case version of the server name.
    :type instance_name:
      :py:obj:`str`
    :param db
      Path to a pre-populated text file to use for the
      database.
      Optional.  Default is to create a new temporary file and populate it
      based on the devices and properties supplied in `devices_info`.
    :type db:
      :py:obj:`str`
    :param host:
      Hostname to use for device server's ORB endpoint.
      Optional.  Default is a local IP address.
    :type host:
      :py:obj:`str`
    :param port:
      Port number to use for the device server's ORB endpoint. 
      Optional.  Default is chosen by omniORB.
    :type port:
      :py:obj:`int`
    :param debug:
      Debug level for the device server logging.
      0=OFF, 1=FATAL, 2=ERROR, 3=WARN, 4=INFO, 5=DEBUG.
      Optional. Default is warn.
    :type debug:
      :py:obj:`int`
    :param process:
      True if the device server should be launched in a new process, otherwise
      use a new thread.  Note:  if the context will be used mutiple times, it
      may seg fault if the thread mode is chosen.
      Optional.  Default is thread.
    :type process:
      :py:obj:`bool`
    :param daemon:
      True if the new thread/process must be created in daemon mode.  
      Optional.  Default is not daemon.
    :type daemon:
      :py:obj:`bool`
    :param timeout:
      How long to wait (seconds) for the device server to start up, and also
      how long to wait on joining the thread/process when stopping. 
      Optional.  Default differs for thread and process modes.
    :type timeout:
      :py:obj:`float`
    """
```

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]:
```suggestion
        """Enter method for context support.

        :return:
          Instance of this test context.  Use `get_device` to get proxy
          access to any of the devices started by this context.
        :rtype:
          :class:`~tango.test_context.MultiDeviceTestContext`
        """
```

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py:430]: OK, we'll leave it returning the whole object.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]:
```suggestion
    """Context to run a single device without a database.

    The difference with respect to
    :class:`~tango.test_context.MultiDeviceTestContext` is that it only
    allows to export a single device.

    Example usage::
        
        from time import sleep

        from tango.server import Device, attribute, command
        from tango.test_context import DeviceTestContext

        class PowerSupply(Device):

            @attribute(dtype=float)
            def voltage(self):
                return 1.23

            @command
            def calibrate(self):
                sleep(0.1)

        def test_calibrate():
            '''Test device calibration and voltage reading.'''
            with DeviceTestContext(PowerSupply, process=True) as proxy:
                proxy.calibrate()
                assert proxy.voltage == 1.23
      
    :param device:
      Device class to be run.
    :type device:
      :class:`~tango.server.Device` or :class:`~tango.DeviceImpl`
    :param device_cls:
      The device class can be provided if using the low-level API.
      Optional.  Not required for high-level API devices, of type
      :class:`~tango.server.Device`.
    :type device_cls:
      :class:`~tango.DeviceClass`

     The rest of the parameters are described in
     :class:`~tango.test_context.MultiDeviceTestContext`.
    """
```

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: As above, suggest this is removed, and class-level docstring is used instead.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]:
```suggestion
        """Enter method for context support.

        :return:
          A device proxy to the device started by this context.
        :rtype:
          :class:`~tango.DeviceProxy`
        """
```

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py]: I'm still wondering about this?

**Review by ajoubertza — CHANGES_REQUESTED** [review]

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py:32]: I just noticed that this is also done in the test_utils module, so please add something similar there:
https://github.com/tango-controls/pytango/blob/492c8752650744b88112bebc450f8497a8b0315b/tango/test_utils.py#L27

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: I agree! Done in 6502e9a4.

Related to this and also to the Codacy error I have done a simple modification of the `DeviceTestContext.get_device_access` API in 9d21b03f. We could also think of deprecating this method and providing `DeviceTestContext.get_my_device_access`, but I don't know if it's worth that?

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:215]: Done in 59906332. I didn't know where to put the Device classes implementations so I placed them inline in the example. Also to make it simple I just used HLAPI and didn't use the properties.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in aa5ffa20

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in f558aac1.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in 43ce4798.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in 894b7122

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py:32]: Done in db5de328

**Comment by reszelaz** [comment]
I don't know how to reply to this: https://github.com/tango-controls/pytango/pull/314#discussion_r339224160 so I generate a new comment here...

What do you exactly mean? Whether the `MultiDeviceTestContext.append_db_file` method name is fine? or maybe its API? Or you reconsider having two methods as it was before https://github.com/tango-controls/pytango/pull/314#discussion_r336793696.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Nice example, thanks.

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py]: > I don't know how to reply to this: #314 (comment) so I generate a new comment here...
>
> What do you exactly mean? Whether the MultiDeviceTestContext.append_db_file method name is fine? or maybe its API? Or you reconsider having two methods as it was before #314 (comment).

I'm thinking about the code that will parse this file.  My question is about the format of the line with `DEVICE` in it.  The current version of your `append_db_file` method will have the first device name on the same line as the `DEVICE` part.  Then the next device is on a new line:
```
Clock/clock/DEVICE/Clock: "test/nodb/clock1"
: "test/nodb/clock2"
```

It is just a bit inconsistent, so I was wondering if it works.  Something like this would be more "consistent" for me:
```
Clock/clock/DEVICE/Clock
: "test/nodb/clock1"
: "test/nodb/clock2"
```

I tried exporting some properties from Jive to see the file format, but the files didn't have a `DEVICE` line.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Is there a need for this flush?  I'd expect exiting the context handler to close the file which will flush any pending writes.

File: tango/test_context.py
Comment (ajoubertza) [inline tango/test_context.py]: Minor point, but it will be slightly easier to read as:
```python
if device_name not in self._devices:
```

**Review by ajoubertza — COMMENTED** [review]
This is looking good.  Just my question about the format of the db file that we need to resolve.

**Comment by ajoubertza** [comment]
> I don't know how to reply to this: [#314 (comment)](https://github.com/tango-controls/pytango/pull/314#discussion_r339224160) so I generate a new comment here...

I'm not sure why, but I only see the reply option for this on the "Files changed" tab, not this "Conversation" tab.

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py]: I'm happy with your workaround for `get_device_access` as is.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: No, I don't remember now how it got there:) I just tested my examples without and it sill works.
I just removed it in 0ade4703.

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Sorry I didn't see your initial comment. Now I understand your concern.
Actually now testing it with two devices of the same class I have realized that it does now work correctly. I will investigate it tomorrow.

**Comment by reszelaz** [comment]
Sorry, I have troubles with replying to your code-review comments. So I reply here for the one about writing in the DB file multiple devices of the same class. I have fixed this in 7c7a6494. In summary, these devices must be written in one line, comma-separated. Now this case works:

```
Pool/pool/DEVICE/Pool: "pool/test/1"
Pool/pool/DEVICE/Controller: "controller/test/1",\
                             "controller/test/2"

#############################################
# CLASS Pool


# CLASS Pool attribute properties


#############################################
# CLASS Controller


# CLASS Controller attribute properties



# DEVICE pool/test/1 properties 


# DEVICE pool/test/1 attribute properties

# DEVICE controller/test/1 properties 

controller/test/1->Id: 1
controller/test/1->Klass: DummyMotorController
controller/test/1->Library: DummyMotorController.py
controller/test/1->Type: Motor

# DEVICE controller/test/1 attribute properties

# DEVICE controller/test/2 properties 

controller/test/2->Id: 2
controller/test/2->Klass: DummyCounterTimerController
controller/test/2->Library: DummyCounterTimerController.py
controller/test/2->Type: CTExpChannel

# DEVICE controller/test/2 attribute properties
```

File: tango/test_context.py
Comment (reszelaz) [reply tango/test_context.py]: Done in d30daad5.

File: tango/test_context.py
Comment (ajoubertza) [reply tango/test_context.py]: The update in 7c7a649 for multiple devices look good.  Thanks.

**Review by ajoubertza — DISMISSED** [review]
Looks good to me!
(Not sure if we can do anything about Codacy's "MultiDeviceTestContext.__init__ is too complex (16)" warning).

I would still like some unit tests for this, but I guess we can add them in a new PR.

**Comment by reszelaz** [comment]
> (Not sure if we can do anything about Codacy's "MultiDeviceTestContext.**init** is too complex (16)" warning).

Strange that it started to fail with a commit which has nothing to do with this `__init__`: 0ade4703.

> I would still like some unit tests for this, but I guess we can add them in a new PR.

I can try to add tests. Where should I place them? in `tests/test_context.py` or in `tests/test_test_context.py`?
Is it fine to just use the context to start DS with some _well known_ device(s) and check if the device is correctly exported and behaves as expected? maybe also try to use properties?

**Comment by ajoubertza** [comment]
> I can try to add tests. Where should I place them? in `tests/test_context.py` or in `tests/test_test_context.py`?

`tests/test_test_context.py` would follow the pattern we are using: 
`tests/test_ + <thing under test>`

> Is it fine to just use the context to start DS with some _well known_ device(s) and check if the device is correctly exported and behaves as expected? maybe also try to use properties?

You can define a few simple devices in that file.  You will see many examples of `TestDevice` classes created as part of each test in in `test_server.py`.  You'll also see the `server_green_mode` fixture which tries all the the different types, you can include that, if you like.

Yes, just test simple behaviour:
- test that some invalid configurations fail
- test multiple devices (the derived DeviceTestContext is already testing single devices) 
- test get_device_access method
- test get_device and that cache works
- maybe set the State in init_device and then check that it is correct
- test a command
- test reading an attribute
- test that property specified at startup works

**Comment by ajoubertza** [comment]
@reszelaz I added some tests.  I'll merge in the next day or so, unless you have a comment.

**Comment by reszelaz** [comment]
Thanks Anton for writing the test. I am a little bit busy this days. Of course go ahead with the merge!

**Review by ajoubertza — APPROVED** [review]

**Pull request merged**
